hcl/parse.go

```diff
@@ -0,0 +1,318 @@
+
+			//break // let's see if another resource even matches
+		}
+	}
+
+	for _, e := range c.Edges {
+		if _, ok := lookup[key{strings.ToLower(e.From.Kind), e.From.Name}]; !ok {
+			return nil, fmt.Errorf("can't find 'from' name")
+		}
+		if _, ok := lookup[key{strings.ToLower(e.To.Kind), e.To.Name}]; !ok {
+			return nil, fmt.Errorf("can't find 'to' name")
+		}
+		from := lookup[key{strings.ToLower(e.From.Kind), e.From.Name}]
+		to := lookup[key{strings.ToLower(e.To.Kind), e.To.Name}]
+		edge := &resources.Edge{
+			Name:   e.Name,
+			Notify: e.Notify,
+		}
+		graph.AddEdge(from, to, edge)
+	}
+
+	return graph, nil
+}
+
+func loadHcl(f *string) (*Config, error) {
+	if f == nil {
+		return nil, fmt.Errorf("empty file given")
+	}
+
+	log.Printf("loading file %s", *f)
+	data, err := ioutil.ReadFile(*f)
+	if err != nil {
+		return nil, fmt.Errorf("unable to read file: %v", err)
+	}
+
+	log.Printf("parsing contents: %s", data)
+	file, err := hcl.ParseBytes(data)
+	if err != nil {
+		return nil, fmt.Errorf("unable to parse file: %s", err)
+	}
+
+	config := new(Config)
+
+	list, ok := file.Node.(*ast.ObjectList)
+	if !ok {
+		return nil, fmt.Errorf("unable to parse file: file does not contain root node object")
+	}
+
+	if resources := list.Filter("resource"); len(resources.Items) > 0 {
+		var err error
+		config.Resources, err = loadResourcesHcl(resources)
+		if err != nil {
+			return nil, fmt.Errorf("unable to parse: %s", err)
+		}
+	}
+
+	if edges := list.Filter("edge"); len(edges.Items) > 0 {
+		var err error
+		config.Edges, err = loadEdgesHcl(edges)
+		if err != nil {
+			return nil, fmt.Errorf("unable to parse: %s", err)
+		}
+	}
+
+	return config, nil
+}
+
+func loadEdgesHcl(list *ast.ObjectList) ([]*Edge, error) {
+	list = list.Children()
+	if len(list.Items) == 0 {
+		return nil, nil
+	}
+
+	var result []*Edge
+
+	for _, item := range list.Items {
+		name := item.Keys[0].Token.Value().(string)
+
+		var config Edge
+		if err := hcl.DecodeObject(&config, item.Val); err != nil {
+			return nil, fmt.Errorf(
+				"Error reading config for %s: %s",
+				name,
+				err)
+		}
+
+		config.Name = name
+
+		result = append(result, &config)
+	}
+	return result, nil
+}
+
+func loadResourcesHcl(list *ast.ObjectList) ([]*Resource, error) {
+	list = list.Children()
+	if len(list.Items) == 0 {
+		return nil, nil
+	}
+
+	var result []*Resource
+
+	log.Printf("HCLParse: parsing %d items", len(list.Items))
+	for _, item := range list.Items {
+		kind := item.Keys[0].Token.Value().(string)
+		name := item.Keys[1].Token.Value().(string)
+
+		var listVal *ast.ObjectList
+		if ot, ok := item.Val.(*ast.ObjectType); ok {
+			listVal = ot.List
+		} else {
+			return nil, fmt.Errorf("module '%s': should be an object", name)
+		}
+
+		res, err := resources.NewEmptyNamedResource(kind)
+		if err != nil {
+			log.Printf("HCLParse: unable to parse resource: %v", err)
+			return nil, err
+		}
+
+		res.SetName(name)
+		res.SetKind(kind)
```

Comment:
No worries :) There are still a bunch of rough spots in the code, since our goal is to get to MVP and it has occasionally let some less than perfect messes in. Also learning experiences, etc...